File: fields/src/test/java/network/aika/fields/oneobject/TestType.java
package network.aika.fields.oneobject;


import network.aika.type.Type;
import network.aika.type.TypeRegistry;
import network.aika.type.relations.Relation;
import network.aika.type.relations.RelationOne;
import network.aika.type.relations.RelationSelf;

import java.util.List;


public class TestType extends Type<TestType, TestObject> {

    public static final RelationSelf<TestType, TestObject> SELF = new RelationSelf<>(0, "TEST_SELF");

    public static final RelationOne<TestType, TestObject, TestType, TestObject> TEST_RELATION_FROM = new RelationOne<>(TestObject::getRelatedTestObject, 1, "TEST_FROM");
    public static final RelationOne<TestType, TestObject, TestType, TestObject> TEST_RELATION_TO = new RelationOne<>(TestObject::getRelatedTestObject, 2, "TEST_TO");

    static {
        TEST_RELATION_TO.setReversed(TEST_RELATION_FROM);
        TEST_RELATION_FROM.setReversed(TEST_RELATION_TO);
    }

    public TestType(TypeRegistry registry, String name) {
        super(registry, name);
    }

    @Override
    public Relation<TestType, TestObject, ?, ?>[] getRelationTypes() {
        return new Relation[] {SELF, TEST_RELATION_FROM, TEST_RELATION_TO};
    }

    public TestObject instantiate() {
        return instantiate(
                List.of(TestType.class),
                List.of(this)
        );
    }
}

File: src/fields/test_type.cpp
#include "fields/test_type.h"
#include "fields/test_object.h"

// Initialize static members
RelationOne TestType::SELF(0, "TEST_SELF");
RelationOne TestType::TEST_RELATION_FROM(1, "TEST_FROM");
RelationOne TestType::TEST_RELATION_TO(2, "TEST_TO");

// Static block equivalent (executed once during initialization)
struct StaticInitializer {
    StaticInitializer() {
        TestType::TEST_RELATION_TO.setReversed(&TestType::TEST_RELATION_FROM);
        TestType::TEST_RELATION_FROM.setReversed(&TestType::TEST_RELATION_TO);
    }
} staticInitializerInstance;

// Constructor
TestType::TestType(TypeRegistry* registry, const std::string& name)
    : Type(registry, name) {
    // Add relations to the list
    relations.push_back(&SELF);
    relations.push_back(&TEST_RELATION_FROM);
    relations.push_back(&TEST_RELATION_TO);
}

// Instantiate method
TestObject* TestType::instantiate() {
    return new TestObject(this);
}
